README.md

```diff
@@ -0,0 +1,486 @@
+const Pair = def('Pair', {}, [a, b, $Pair(a, b)], (x, y) => ({
+  '0': x,
+  '1': y,
+  '@@type': 'sanctuary-def/Pair',
+  'length': 2,
+  'toString': () => 'Pair(' + JSON.stringify(x) + ', ' + JSON.stringify(y) + ')',
+}));
+
+//    Rank :: Type
+const Rank = $.NullaryType(
+  'sanctuary-def/Rank',
+  x => typeof x === 'string' && /^([A23456789JQK]|10)$/.test(x),
+  'A'
+);
+
+//    Suit :: Type
+const Suit = $.NullaryType(
+  'sanctuary-def/Suit',
+  x => typeof x === 'string' && /^[\u2660\u2663\u2665\u2666]$/.test(x),
+  '\u2660'
+);
+
+//    Card :: Pair Rank Suit
+const Card = $Pair(Rank, Suit);
+
+//    showCard :: Card -> String
+const showCard =
+def('showCard', {}, [Card, $.String], card => card[0] + card[1]);
+
+showCard(Pair('A', '♠'));
+// => 'A♠'
+
+showCard(Pair('X', '♠'));
+// ! TypeError: ‘showCard’ expected a value of type (Pair Rank Suit) as its first argument; received Pair("X", "♠")
+```
+
+#### `RecordType`
+
+`RecordType` is used to construct record types. The type definition specifies
+the name and type of each required field.
+
+To define a record type one must provide:
+
+  - an object mapping field name to type.
+
+```haskell
+RecordType :: {Type} -> Type
+```
+
+For example:
+
+```javascript
+//    FiniteNumber :: Type
+const FiniteNumber = $.NullaryType(
+  'sanctuary-def/FiniteNumber',
+  x => Object.prototype.toString.call(x) === '[object Number]' && isFinite(x)
+);
+
+//    Point :: Type
+const Point = $.RecordType({x: FiniteNumber, y: FiniteNumber});
+
+//    dist :: Point -> Point -> FiniteNumber
+const dist =
+def('dist', {}, [Point, Point, FiniteNumber],
+    (p, q) => Math.sqrt(Math.pow(p.x - q.x, 2) + Math.pow(p.y - q.y, 2)));
+
+dist({x: 0, y: 0}, {x: 3, y: 4});
+// => 5
+
+dist({x: 0, y: 0}, {x: 3, y: 4, color: 'red'});
+// => 5
+
+dist({x: '0', y: '0'}, {x: '3', y: '4'});
+// ! TypeError: ‘dist’ expected a value of type { x :: FiniteNumber, y :: FiniteNumber } as its first argument; received {"x": "0", "y": "0"}
+
+dist(0);
+// ! TypeError: ‘dist’ expected a value of type { x :: FiniteNumber, y :: FiniteNumber } as its first argument; received 0
+```
+
+### Type classes
+
+`concatS`, defined earlier, is a function which concatenates two strings.
+This is overly restrictive, since other types support concatenation (`Array`,
+for example).
+
+One could use a type variable to define a polymorphic "concat" function:
+
+```javascript
+//    _concat :: a -> a -> a
+const _concat =
+def('_concat', {}, [a, a, a], (x, y) => x.concat(y));
+
+_concat('fizz', 'buzz');
+// => 'fizzbuzz'
+
+_concat([1, 2], [3, 4]);
+// => [1, 2, 3, 4]
+
+_concat([1, 2], 'buzz');
+// ! TypeError: ‘_concat’ expected a value of type (Array Number) as its second argument; received "buzz"
+```
+
+The type of `_concat` is misleading: it suggests that it can operate on any
+two values of *any* one type. In fact there's an implicit constraint, since
+the type must support concatenation (in [Fantasy Land][2] terms, the type
```

Comment:
Since `Semigroup` is already a link to the appropriate part of the FantasyLand specification, might it not be better to say

>    (in [mathematical][2] terms, the type must have a [semigroup][3])
>      [2]: https://en.wikipedia.org/wiki/Semigroup
>      [3]: https://github.com/fantasyland/fantasy-land#semigroup

?